[VideoDatabase] Fix clean database deletes archive:// paths and refactor for bluray:// paths

78andyp opened the pull request:
Problem: Running a database clean deletes archive:// entries even though the actual archive is still present
Fix: archive:// paths need to be resolved to the actual archive file on disk.

Also, as part of investigating the above, I've properly implemented the same for bluray:// paths.

@garbear - should be straightforward. This was something I identified in #25140.

## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP commented:
While this fix achieves the result for a zip that only contains one file at root level, it may be the archive addon's responsibility to check correctly when invoked by the existing line `exists = CDirectory::Exists(path, false);`

Archives can contain multiple levels of subdirectories and their continued existence cannot be checked without the addon's involvement. Checking the root is a degenerate trivial case.

garbear commented:
I was waiting to comment for this exact reason. And only because of my research due to https://github.com/xbmc/xbmc/pull/26227, where I learned about archive:// paths. The fix here seems brittle, and would be better addressed in `CDirectory::Exists()`, if that's indeed the way.

78andyp committed:
Remove special case for bluray:// files from CleanDatabase()

78andyp committed:
Fix CUtil::GetMatchingSources() so bluray:// paths are resolved to their underlying host/base path on disc

78andyp committed:
Fix CUtil::GetMatchingSources() for archive:// (and any other VFS addon) paths.

Resolve these vfs paths to their host path before verifying their source still exists.

This fixes the underlying issue of archive:// paths being cleaned from the database.

78andyp commented:
> I was waiting to comment for this exact reason. And only because of my research due to #26227, where I learned about archive:// paths. The fix here seems brittle, and would be better addressed in `CDirectory::Exists()`, if that's indeed the way.

Thanks. Actually GetDirectory() and CFile::Exists() work ok for archive:// files.
The issue is properly identifying the source which is done in CUtil::GetMatchingSources() - once that's ok, everything else works.

78andyp commented:
> While this fix achieves the result for a zip that only contains one file at root level, it may be the archive addon's responsibility to check correctly when invoked by the existing line `exists = CDirectory::Exists(path, false);`
> 
> Archives can contain multiple levels of subdirectories and their continued existence cannot be checked without the addon's involvement. Checking the root is a degenerate trivial case.

Thanks. I think what i've implemented now is a better fix.

Along the way I also realised that the previous fix for bluray:// paths in library cleaning was also 'a degenerate trivial case' so I've properly implimented CFile::Exists() for bluray:// files.

garbear commented on xbmc/Util.cpp:
```diff
@@ -1251,6 +1251,21 @@ int CUtil::GetMatchingSource(const std::string& strPath1,
   if (checkURL.IsProtocol("multipath"))
     strPath = CMultiPathDirectory::GetFirstPath(strPath);
 
+  // Check to see if protocol is VFS addon (eg. archive://)
+  if (!checkURL.GetProtocol().empty() && CServiceBroker::IsAddonInterfaceUp())
+  {
+    for (const auto& vfsAddon : CServiceBroker::GetVFSAddonCache().GetAddonInstances())
+    {
+      auto prots = StringUtils::Split(vfsAddon->GetProtocols(), "|");
+      if (vfsAddon->HasDirectories() &&
+          std::ranges::find(prots, checkURL.GetProtocol()) != prots.end())
+      {
+        strPath = checkURL.GetHostName(); // resolve host (eg. actual .zip file)
+        break;
+      }
+    }
+  }
+
```
Change looks correct. But I feel I've written VFS addon checks myself that looked similar to this. Can we factor out all these checks? As it sure seems like a lot of add-on logic for a generic Util.cpp file.

garbear replied:
Actually, I was thinking of the `CServiceBroker::GetVFSAddonCache().GetAddonInstances()` add-on logic affected by https://github.com/xbmc/xbmc/pull/26224. Use your judgement if anything can be combined.

78andyp replied:
Looking at the comments for #26566 I wonder whether a lot of this can be simplified with `IDirectory::Resolve()`

garbear commented on xbmc/Util.cpp:
```diff
@@ -1251,6 +1251,21 @@ int CUtil::GetMatchingSource(const std::string& strPath1,
+      if (vfsAddon->HasDirectories() &&
+          std::ranges::find(prots, checkURL.GetProtocol()) != prots.end())
+      {
+        strPath = checkURL.GetHostName(); // resolve host (eg. actual .zip file)
```
I prefer short lines to help developers with large fonts or small screens. We don't pay by the line, so something like:

```c++

      {
        // Resolve host (eg. actual .zip file)
        strPath = checkURL.GetHostName();
        break;
      }
```

Also, as you saw, comments being sentences, I like Sentence casing too for code clarity.

garbear commented:
None of the fixes looked incorrect. One commit had really good commit description of the "why" that helped scrutinize the solution. One comment was the "what" without a "why", so I went on faith.

CrystalP commented:
Is the archive problem still occuring? Looks like I was able to repro in Jan but with recent master I am not able to, with a movie in a .zip file after having downloaded the archive add-on. The movie, file and path remain in the db after a clean library.

For Bluray, code from CVideoDatabase::CleanDatabase() was removed and equivalent (but more complex) code was added to CUtil::GetMatchingSource()
GetDiscBase does more than GetBlurayFile and actually systematically calls GetBlurayFile, and GetMatchingSource is called from more places. What is the benefit?
